Add determinate download progress indicator to samples-lib

shubham7109 opened the pull request:
## Description
PR update the samples`DownloaderActivity.kt` to use the `ArcGISEnvironment.arcGISHttpClient.download` API. This will allow all samples that require file provisioning, to download data with a progress indicator.  

## Links and Data
Sample Epic: `runtime/kotlin/issues/1995`

## What To Review
- Changed files

## How to Test
Run a sample which requires file provisioning:
- Display Dimensions (.mmpk)
- Add feature layers (.gpkg, .geodatabase, .shp)
- Find route in transport network (.zip extracted)

gunt0001 commented on samples-lib/src/main/java/com/esri/arcgismaps/sample/sampleslib/DownloaderActivity.kt:
```diff
@@ -224,43 +233,38 @@ abstract class DownloaderActivity : AppCompatActivity() {
+                if (totalBytes != null) {
+                    // update the download percentage progress
+                    val percentage = ((100.0 * bytesRead) / totalBytes).roundToInt()
+                    Handler(Looper.getMainLooper()).post {
```
Interesting. This should not be necessary. I wonder if we should make the progress listener lambda a suspend function, in which case you could just call something like:
```kotlin
withContext(Dispatchers.Main) {
   progressIndicator.progress = percentage
   progressTV.text = "$percentage%"
}
```

Alternativey we could make sure that the progress listener is always invoked on the main dispatcher.

@changanxian - thoughts?

shubham7109 replied:
Updated this to use your suggestion. This is required since changes to the linear progress indicator is required to run on the main thread:
```kotlin
lifecycleScope.launch(Dispatchers.Main) {
    progressIndicator.progress = percentage
    progressTV.text = "$percentage%"
}

changanxian replied:
Since the listener is updating the progress UI, it needs to run on the `main` thread.  Changing the listener be `suspend` function makes more sense to me.
@gunt0001

gunt0001 replied:
In my mind, invoking the progress listener on the main dispatcher would be evn more convenient, as the user would always be called back on the main dispatcher, so non need at all for any thread switching in user code.

gunt0001 commented on samples-lib/src/main/java/com/esri/arcgismaps/sample/sampleslib/DownloaderActivity.kt:
```diff
@@ -224,43 +233,38 @@ abstract class DownloaderActivity : AppCompatActivity() {
                 return@flow
             }
+            // set up the file at the download path
+            val destinationFilePath = provisionLocation.path + File.separator + portalItem.name
```
There are better options to create this path:
1. Use the `File` API - `val destinationFilePath = File(provisionLocation.path, portalItem.name).toString()`
2. ``val destinationFilePath = "${provisionLocation.path}File.separator${portalItem.name}"`

gunt0001 commented on samples-lib/src/main/java/com/esri/arcgismaps/sample/sampleslib/DownloaderActivity.kt:
```diff
@@ -224,43 +233,38 @@ abstract class DownloaderActivity : AppCompatActivity() {
+                    val provisionFile = File(destinationFilePath)
+                    // set up the zip input stream
                     val fileInputStream = FileInputStream(destinationFilePath)
                     val zipInputStream = ZipInputStream(BufferedInputStream(fileInputStream))
```
I believe all this unzipping logic should be implemented in a "try with resources" style, i.e. the input streams should be closed in a `finally` block. In Kotlin you would typically do this with the extension function [Closable.use](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html).

shubham7109 replied:
I've updated the input streams to:
```kotlin
FileInputStream(destinationFilePath).use { fileInputStream ->
    ZipInputStream(BufferedInputStream(fileInputStream)).use { zipInputStream ->
        ...
    }
}
```

shubham7109 commented:
Thanks for the review @gunt0001, added the feedback. Ready for your re-review.

changanxian commented on samples-lib/src/main/java/com/esri/arcgismaps/sample/sampleslib/DownloaderActivity.kt:
```diff
@@ -224,59 +231,56 @@ abstract class DownloaderActivity : AppCompatActivity() {
+            val downloadURL =
+                "${portalItem.portal.url}/sharing/rest/content/items/${portalItem.itemId}/data"
             // get the data of the PortalItem
-            val portalItemData = portalItem.fetchData()
```
Did you create an issue for this? we need to investigate why the usage fails.

shubham7109 replied:
The parent issue of this PR (`kotlin/issues/1995`) was created when we noticed `fetchData()` will download all the data into memory, so it was unsuitable for large file downloads. 

Should there be an individual issue to update the `fetchData()` design?

gunt0001 commented on samples-lib/src/main/java/com/esri/arcgismaps/sample/sampleslib/DownloaderActivity.kt:
```diff
@@ -224,59 +231,56 @@ abstract class DownloaderActivity : AppCompatActivity() {
                 return@flow
             }
+            // set up the file at the download path
+            val destinationFilePath = File(provisionLocation.path, portalItem.name).toString()
```
Since you require a type `File` further down in your code, you might as well hold on to a `File` instance instead of `String`:
```suggestion
            val destinationFilePath = File(provisionLocation.path, portalItem.name)
```

gunt0001 commented on samples-lib/src/main/java/com/esri/arcgismaps/sample/sampleslib/DownloaderActivity.kt:
```diff
@@ -224,59 +231,56 @@ abstract class DownloaderActivity : AppCompatActivity() {
-                    downloadCount = byteArrayInputStream.read(data)
+            ArcGISEnvironment.arcGISHttpClient.download(
+                url = downloadURL,
+                destinationFile = File(destinationFilePath)
```
With my comment above this could be changed to:
```suggestion
                destinationFile = destinationFilePath
```

gunt0001 commented on samples-lib/src/main/java/com/esri/arcgismaps/sample/sampleslib/DownloaderActivity.kt:
```diff
@@ -224,59 +231,56 @@ abstract class DownloaderActivity : AppCompatActivity() {
+                if (totalBytes != null) {
+                    // update the download percentage progress
+                    val percentage = ((100.0 * bytesRead) / totalBytes).roundToInt()
+                    lifecycleScope.launch(Dispatchers.Main) {
```
This is better 👍

gunt0001 commented on samples-lib/src/main/java/com/esri/arcgismaps/sample/sampleslib/DownloaderActivity.kt:
```diff
@@ -224,59 +231,56 @@ abstract class DownloaderActivity : AppCompatActivity() {
+                                    File(provisionLocation.path, zipEntry.name).mkdirs()
+                                } else {
+                                    val file = File(provisionLocation.path, zipEntry.name)
+                                    val fout = FileOutputStream(file)
```
Should also make use of `use` here:
```suggestion
                                    FileOutputStream(file).use { fileOutputStream ->
```

shubham7109 replied:
Yeah, it should! Also removed the `fout.close()` since `.use` should be a close the output stream.

gunt0001 commented on samples-lib/src/main/java/com/esri/arcgismaps/sample/sampleslib/DownloaderActivity.kt:
```diff
@@ -224,59 +231,56 @@ abstract class DownloaderActivity : AppCompatActivity() {
+                                zipEntry = zipInputStream.nextEntry
                             }
-                            fout.close()
+                            zipInputStream.close()
```
This call should not be necessary, since the `use { }` block will take care of it.

shubham7109 commented:
Thanks for the review @gunt0001 again. Hopefully, it looks good.

gunt0001 reviewed (approved):
@shubham7109 - looks good 👍

changanxian commented on samples-lib/src/main/java/com/esri/arcgismaps/sample/sampleslib/DownloaderActivity.kt:
```diff
@@ -224,59 +231,53 @@ abstract class DownloaderActivity : AppCompatActivity() {
+                url = downloadURL,
+                destinationFile = destinationFile
+            ) { totalBytes, bytesRead ->
+                if (totalBytes != null) {
```
if `totalBytes` is null, the ` progressIndicator` will be not shown, but downloading may still happen, since the server may not provide the property.  It is suggested to show a UX Spinner since we don't know how much in progress.

changanxian replied:
maybe you could keep the old spinner to show if the `totalBytes` is null.

changanxian commented:
@shubham7109 looks good, one comment.

shubham7109 commented:
Added PR feedback, and updated with `v.next` branch. Added code to modify the spinner visibility using `Dispatchers.Main`.  @changanxian Ready for your review.

changanxian commented:
Looks good
